File: sources/opencv-samples/9_ImagesSubtraction/main.cpp
// https://docs.opencv.org/4.x/db/df5/tutorial_linux_gcc_cmake.html

#include <stdio.h>
#include <opencv2/opencv.hpp>
using namespace cv;
using namespace std;

int main(int argc, char** argv)
{
    if ( argc < 3 || argc > 4 )
    {
        printf("Usage: ./main.out <image1_path> <image2_path> [g|G]\n");
        printf("Read the README.md\n");
        return -1;
    }
    cv::ImreadModes mode = IMREAD_UNCHANGED;
    if (argc == 4)
    {   
        if (!strcmp(argv[3], "g") || !strcmp(argv[3], "G"))
        { 
            mode = IMREAD_GRAYSCALE;
        }
        else
        {
            printf("Unexpected mode value\n");
            printf("Read the README.md\n");
            return -1;
        }
    }
    Mat image1;
    Mat image2;
    Mat subtracted;

    image1 = imread(argv[1], mode);
    image2 = imread(argv[2], mode);

    if ( !image1.data )
    {
        cout << "Could not read the image " << argv[1] << endl;
        return -1;
    }
    if ( !image2.data )
    {
        cout << "Could not read the image " << argv[2] << endl;
        return -1;
    }

    int channelsImage1 = image1.channels();
    int rowsImage1 = image1.rows;
    int colsImage1 = image1.cols;
    int depthImage1 = image1.depth();

    int channelsImage2 = image2.channels();
    int rowsImage2 = image2.rows;
    int colsImage2 = image2.cols;
    int depthImage2 = image2.depth();

    if (channelsImage1 != channelsImage2 || rowsImage1 != rowsImage2 || colsImage1 != colsImage2 || depthImage1 != depthImage2)
    {
        cout << "Not the same dimensions or depth!\t";
        cout << "image 1: (" << rowsImage1 << "," << colsImage1 << "," << channelsImage1 << ") (" << depthImage1 << ")\t";
        cout << "image 2: (" << rowsImage2 << "," << colsImage2 << "," << channelsImage2 << ") (" << depthImage2 << ")" << endl;
        return -1;
    }

    double delay = (double)getTickCount();
    cv::subtract(image1, image2, subtracted);
    delay = 1000 * ((double)getTickCount() - delay) / getTickFrequency();
    cout << "Subtraction delay: " << (int)(delay * 1000) / 1000.0 << " ms" << endl;

    imshow("Image 1", image1);
    imshow("Image 2", image2);
    imshow("Substract", subtracted);

    std::cout << "Press a key to quit" << std::endl;
    waitKey(0);
    destroyAllWindows();
    return 0;
}

File: sources/opencv-samples/9_ImagesSubtraction/main.py
# https://docs.opencv.org/4.x/db/deb/tutorial_display_image.html

import cv2 as cv
import argparse
from time import time


def main():

    args = parser.parse_args()

    path1 = cv.samples.findFile(args.img1_path)
    path2 = cv.samples.findFile(args.img2_path)

    if not path1:
        sys.exit(f"Image {args.img1_path} not found")
    if not path2:
        sys.exit(f"Image {args.img2_path} not found")

    mode = cv.IMREAD_UNCHANGED
    if args.grayscale:
        mode = cv.IMREAD_GRAYSCALE
    img1 = cv.imread(path1, mode)
    img2 = cv.imread(path2, mode)

    if img1 is None:
        exit(f"Could not read image {path1}")
    if img2 is None:
        exit(f"Could not read image {path2}")

    if img1.shape != img2.shape or img1.dtype != img2.dtype:
        exit(f"Not the same dimensions or data type!\timage 1: {img1.shape} ({img1.dtype})\timage 2: {img2.shape} ({img2.dtype})")

    delay = time()
    subtracted = cv.subtract(img1, img2)
    delay = 1000 * (time() - delay)
    print(f"Subtraction delay: { round(delay * 1000) / 1000} ms")

    cv.imshow("Image 1", img1)
    cv.imshow("Image 2", img2)
    cv.imshow("Subtract", subtracted)
    print("Press a key to quit")
    cv.waitKey(0)
    cv.destroyAllWindows()


if __name__ == '__main__':
    
    parser = argparse.ArgumentParser(description="Images Subtraction main.py CLI")
    parser.add_argument("img1_path", type=str, help="image 1 path to open")
    parser.add_argument("img2_path", type=str, help="image 2 path to open")
    parser.add_argument("-g", "--grayscale", action="store_true", help="read the images with GRAYSCALE")

    main()
